include/ginkgo/core/preconditioner/schwarz.hpp

```diff
@@ -0,0 +1,234 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_PRECONDITIONER_SCHWARZ_HPP_
+#define GKO_PUBLIC_CORE_PRECONDITIONER_SCHWARZ_HPP_
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/diagonal.hpp>
+#include <ginkgo/core/multigrid/amgx_pgm.hpp>
+#include <ginkgo/core/multigrid/multigrid_level.hpp>
+
+
+namespace gko {
+/**
+ * @brief The Preconditioner namespace.
+ *
+ * @ingroup precond
+ */
+namespace preconditioner {
+
+
+/**
+ * A Schwarz preconditioner is a simple domain decomposition preconditioner that
+ * generalizes the Block Jacobi preconditioner, incorporating options for
+ * different local subdomain solvers and overlaps between the subdomains.
+ *
+ * See Iterative Methods for Sparse Linear Systems (Y. Saad) for a general
+ * treatment and variations of the method.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  integral type of the preconditioner
+ *
+ * @ingroup schwarz
+ * @ingroup precond
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Schwarz : public EnableLinOp<Schwarz<ValueType, IndexType>>,
+                public WritableToMatrixData<ValueType, IndexType>,
+                public Transposable {
+    friend class EnableLinOp<Schwarz>;
+    friend class EnablePolymorphicObject<Schwarz, LinOp>;
+
+public:
+    using EnableLinOp<Schwarz>::convert_to;
+    using EnableLinOp<Schwarz>::move_to;
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using mat_data = matrix_data<ValueType, IndexType>;
+    using transposed_type = Schwarz<ValueType, IndexType>;
+
+    /**
+     * Returns the number of blocks of the operator.
+     *
+     * @return the number of blocks of the operator
+     */
+    size_type get_num_subdomains() const noexcept { return num_subdomains_; }
+
+    void write(mat_data& data) const override;
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    /**
+     * Returns the subdomain matrices.
+     *
+     * @return the subdomain matrices
+     */
+    std::vector<std::shared_ptr<LinOp>> get_subdomain_matrices() const
+    {
+        return subdomain_matrices_;
+    }
+
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Array of subdomain sizes.
+         */
+        std::vector<size_type> GKO_FACTORY_PARAMETER_VECTOR(
+            subdomain_sizes, std::vector<size_type>{});
```

Comment:
structural blocks may work well work block Jacobi, but I don't think this maps well to larger domains, and even less so to overlaps (requiring explicit reordering to make it work generically). I think we probably need to integrate the Partition and corresponding kernels somehow.

Reply:
Yes, I agree that for larger subdomains structural blocks might not be ideal, but I think for a single executor implementation, it is a good start. For single executor implementation, you can use an index_set or partition as you mention to construct non-contiguous blocks, but the problem with those is that you have to deal with a lot of complications: map rhs and solution vector to correct indices/maintain copies and copy back to the vector. So in general, it might be better to reorder the system matrix before hand and use contiguous blocks.